### PR TITLE
indexed columns are utf8 and table creation check

### PR DESCRIPTION
**Pull request by jnorell**
use utf8 charset for indexed columns, and test existence of table to see if creation is needed

See https://wordpress.org/support/topic/export-personal-data-error/

### CODE REVIEW

**Comment by jnorell** [comment]
The explicit check if the table exists is because create_table is only called when the version number is less than current, and my version number was at 2 even though the table failed to create.  Probably the version number (pum_subscribers_db_version) should not be updated if the table create fails.

I'd suggest you might even want to move the update_option() to the parent's __construct(), assuming that is likely behavior wanted in all classes extending PUM_Abstract_Database, and also because __construct() is what actually does the version comparison of that field.

I'd be glad to tweak things and resubmit, though as a heads up I have more time towards the end of the week, so possible I won't get to it timely enough for your needs.

Thanks

**Comment by jnorell** [comment]
Note this latest commit is untested, but to demonstrate what I suggested above.

**Comment by danieliser** [comment]
Sorry. To clarify the table install failed previously for whatever reason due to the indexes. If we solve that, no further need to check if table exists if that makes since. This would essentially add a needless query on every call to the dB.

Instead if we focus on ensuring the query to create the table is fixed then bump the dB version it will back fill for you and others that failed to install correctly the first time.

In short yours only broke because of MySQL error on table create. Fix that issue and no need for the other bandaids if that makes since.

**Comment by danieliser** [comment]
I like the second version of testing right after creation for the existence of the table. Clean solution without any further extra queries unless table wasn't created.

That said been looking and I'm confused as to how this happens in general. To be clear we use the existing collation and charset that all other tables use in your DB. IE the DB default. If the table is getting installed with utf8mb4 it means that the entire DB is utf8mb4. If that is the case then the error doesn't make since. One of the recommended solutions in the stackoverflow ticket you linked is to set the DB to match the table which fixes the index sizing issue.

In short if your DB & table mismatch the error could occur, but that means that $wpdb is returning the wrong thing, something your host may be able to answer.

If they don't mismatch this shouldn't be occurring.

Lastly, my local DB is utf8mb4, the table is utf8mb4 and no errors occurred during creation of the table just now when I tested again. Meaning its either a very specific issue with a specific version of MySQL or your DB charset is wonky, to begin with, and we need to know why.

From what I can tell if you set the individual columns to utf8 and that worked, it likely means your DB has default set to utf8 as well, but told WP to use the other.

Look forward to hearing your thoughts. Will merge the db_ver option update mechanism as I think that makes a lot of since.

**Comment by danieliser** [comment]
I want to add that I just checked and my local DB has default collation of utf8_general_ci, but all tables and columns have utf8mb4_unicode_ci

Not sure but this may be something in core, and we would actually be creating new issues for users by changing it for one column.

May be simpler to just remove the indexes or reducing them to 191 characters rather than 255.

The indexes are likely un-neccessary optimizations.

**Comment by danieliser** [comment]
Just realized, WP doesn't create the DB. We do or the app/host etc does.

So the utf8_general_ci comes from that likely. WP itself installs all the tables using utf8mb4_unicode_ci

With that in mind, I deleted the table and manually ran the query, which though successful did give the warning x2: Warning: #1071 Specified key was too long; max key length is 767 bytes

So really this is a WP core issue we need to add to Trac, but I think the solution is to remove the indexes for now or shorten the char length to 191, testing that length now to see what that looks like and if the hashes store properly.

**Comment by danieliser** [comment]
Can confirm the length of 191 resolves the error. Hash is only 32 chars so can shorten that entirely.

So do this, clean up the patch to only change the length of email to 191, and email_hash to 32, as well as moving the db_ver like you did. Will get it merged for an update I already have about ready.

**Comment by danieliser** [comment]
@jnorell I went ahead and pushed the changes to master per this PR. Noticed just now you said you may not be available til next week. Added a note in the readme for the assist.

**Comment by danieliser** [comment]
Much appreciate the assist. Feel free to submit PRs any time ;)
